### PR TITLE
[BACKPORT] Cache entry event improvements

### PR DESCRIPTION
**Pull request by serkan-ozal**
Before this improvement;

- Even though there is no cache entry listener registered and only invalidation listener is registered, cache entry events are created and sent to event system because of registered invalidation listener. By this improvement, cache entry events are sent to event system if there is any cache entry listener registered. 

- Same logic is available also for cache invalidation listener

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 1]
```diff
@@ -52,11 +51,19 @@
         implements ICacheService, PostJoinAwareService {
 
     protected final ConcurrentMap<String, CacheConfig> configs = new ConcurrentHashMap<String, CacheConfig>();
+    protected final ConcurrentMap<String, CacheContext> cacheContexes = new ConcurrentHashMap<String, CacheContext>();
```
Comment (gurbuzali) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: `cacheContexts` ?

File: hazelcast/src/main/java/com/hazelcast/cache/impl/CacheService.java [hunk 2]
```diff
@@ -239,4 +240,86 @@ public void run() {
...
+            // We don't support this at the moment, because it is not used at the moment
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void clear() {
+            // We don't support this at the moment, because
+            //   - It is not used at the moment
+            //   - It may or may not use "poll" method internally and this depends on the implementation
+            //     so it may change between different version of Java
```
Comment (gurbuzali) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/CacheService.java]: missing `throw new UnsupportedOperationException();` ?

Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/CacheService.java]: yes, my bad. good point :+1:

**Comment by gurbuzali** [comment]
some minor comments, overall good :+1:

**Comment by gurbuzali** [comment]
I should add that some tests related with the newly introduced `NotifiableEventListener` would be good

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java (hunk 1 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: right, my typo wrong

**Pull request merged**